### PR TITLE
Use webpage to to set configuration, fixes #140

### LINKED ISSUE #140: Settings via web GUI
The settings in userdefines.h should be adjustable via web interface. At least a part of it.
Started to work on that.

### PR DESCRIPTION
**Pull request by rexfue**
lorawan.cpp:
 - only compile, if CPU==STICK, i.e if we have LoRa Hardware
 - fetch LoRa credentials from config variables

multigeiger.ino:
 - use config variables instead of userdefines

transmission.cpp:
 - replace SEND2LORA by CPU==STICK
 - use config variables to decide where to send
 - send to LoRa only, if LoRa credentials are known

webconf.cpp
 - add all config variable to web page
 - do restart after saving config

userdefines-example.h
 - remove now unnecessary #defines

### CODE REVIEW

File: multigeiger/webconf.cpp [hunk 1]
```diff
@@ -5,8 +5,47 @@
...
+// WiFi hardware: send to wifi servers and not to LoRa
+char sendToLora[2] = "0";
+char sendToCommunity[2] = "1";
+char sendToMadavi[2] = "1";
+#endif
+
+IotWebConfSeparator sep0 = IotWebConfSeparator("Tick settings");
+IotWebConfParameter startSoundParam = IotWebConfParameter("Start sound (1=ON)", "startSound", startSound, 2, "number", "0/1", NULL, "min='0' max='1' step='1'");
+IotWebConfParameter speakerTickParam = IotWebConfParameter("Speaker tick (1=ON)", "speakerTick", speakerTick, 2, "number", "0/1", NULL, "min='0' max='1' step='1'");
+IotWebConfParameter ledTickParam = IotWebConfParameter("LED ticker (1=ON)", "ledTick", ledTick, 2, "number", "0/1", NULL, "min='0' max='1' step='1'");
```
Comment (ThomasWaldmann) [inline multigeiger/webconf.cpp]: LED tick
(consistency)

File: multigeiger/webconf.cpp [hunk 2]
```diff
@@ -90,3 +149,11 @@ void setup_webconf() {
   });
 }
 
+// Check value of settings parameters
+boolean parameterTrue(char *parameter) {
+  return (strcmp(parameter, "1") == 0);
+}
```
Comment (ThomasWaldmann) [inline multigeiger/webconf.cpp]: don't call this from all over the place, but rather use bool variables.

File: multigeiger/webconf.h [hunk 3]
```diff
@@ -6,9 +6,40 @@
...
+extern IotWebConfSeparator sep1;
+extern IotWebConfParameter sendToCommunityParam;
+extern IotWebConfParameter sendToMadaviParam;
+#if CPU==STICK
+extern IotWebConfSeparator sep3;
+extern IotWebConfParameter sendToLoraParam;
+extern IotWebConfParameter appeuiParam;
+extern IotWebConfParameter deveuiParam;
+extern IotWebConfParameter appkeyParam;
+#endif
```
Comment (ThomasWaldmann) [inline multigeiger/webconf.h]: why are these IotWebConfParameter here?

File: multigeiger/webconf.cpp [hunk 4]
```diff
@@ -80,6 +120,25 @@ void setup_webconf() {
   iotWebConf.getWifiSsidParameter()->label = "WiFi client SSID";
   iotWebConf.getWifiPasswordParameter()->label = "WiFi client password";
 
+  // add the setting parameter
```
Comment (ThomasWaldmann) [inline multigeiger/webconf.cpp]: maybe rather omit such comments, that's quite obvious from the code already.

File: multigeiger/webconf.cpp [hunk 5]
```diff
@@ -80,6 +120,25 @@ void setup_webconf() {
...
+  iotWebConf.addParameter(&sep0);
+  //iotWebConf.addParameter(&startSoundParam1);
```
Comment (ThomasWaldmann) [inline multigeiger/webconf.cpp]: remove?

File: multigeiger/loraWan.cpp [hunk 6]
```diff
@@ -293,4 +283,38 @@ transmissionStatus_t lorawan_send(uint8_t txPort, uint8_t *txBuffer, uint8_t txS
     }
   }
 }
+
+//convert hexstring to len bytes of data
+//returns 0 on success, -1 on error
+//data is a buffer of at least len bytes
+//hexstring is upper or lower case hexadecimal, NOT prepended with "0x"
+int hex2data(unsigned char *data, const char *hexstring, unsigned int len, bool reverse) {
```
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/loraWan.cpp]: hmm, stdlib stuff likely can't support reverse. can this be solved differently or separately?

File: multigeiger/multigeiger.ino [hunk 7]
```diff
@@ -66,7 +66,7 @@ void setup() {
   setup_webconf();
   setup_transmission(VERSION_STR, ssid);
 
-  if (PLAY_SOUND)
+  if (parameterTrue(startSound))
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: calling parameterTrue() is slightly ugly and always causing overhead. use a bool variable.

File: multigeiger/webconf.cpp [hunk 8]
```diff
@@ -5,8 +5,47 @@
...
-
-#define CONFIG_VERSION "012"  // for IoTWebConfig
+#include "userdefines.h"
+
+// All these flags are ON (=1) at startup
+// They all can be changed at any time using the web interface
+char speakerTick[2] = "1";
+char startSound[2] = "1";
+char ledTick[2] = "1";
+char showDisplay[2] = "1";
```
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/webconf.cpp]: BTW, this is where / why you still need SPEAKER_TICK value if you do not want to just hardcode "1" (true) here.

File: multigeiger/webconf.cpp [hunk 9]
```diff
@@ -67,7 +106,8 @@ void handleRoot(void) {  // Handle web requests to "/" path.
 }
 
 void configSaved(void) {
-  log(INFO, "Config saved");
+  log(INFO, "Config saved .. restarting");
+  ESP.restart();
```
Comment (ThomasWaldmann) [inline multigeiger/webconf.cpp]: maybe we could detect actual config change somehow? maybe supported by iotWebConf already?

and only restart if changed?

File: multigeiger/loraWan.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (rexfue) [reply multigeiger/loraWan.cpp]: Separately sure, differently may be. Do you have a good solution?

File: multigeiger/webconf.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (rexfue) [reply multigeiger/webconf.cpp]: To use a 'load_config()_ is a good idea. 
But I still don't see the advantage of a #define SPEAKER_TICK. You have to set the config variables to the the value of the #define at startup anyway.

File: multigeiger/webconf.h (hunk 3 above)
Comment (rexfue) [reply multigeiger/webconf.h]: OK, i see, this is not needed

File: multigeiger/webconf.cpp (hunk 4 above)
Comment (rexfue) [reply multigeiger/webconf.cpp]: Better more than too little comment ...

File: multigeiger/webconf.cpp (hunk 5 above)
Comment (rexfue) [reply multigeiger/webconf.cpp]: OK

File: multigeiger/multigeiger.ino (hunk 7 above)
Comment (rexfue) [reply multigeiger/multigeiger.ino]: I think that results in an unnecessary duplication. You have to copy the bool variables to the config variables and back again.

File: multigeiger/webconf.cpp (hunk 9 above)
Comment (rexfue) [reply multigeiger/webconf.cpp]: Didn't find a way to read the stored configs back from iotwebconf. May be you find one.

File: multigeiger/loraWan.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/loraWan.cpp]: hmm, crap, scanf %x is hex string -> integer... (not byte sequence).

i miss python. :-)

[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/loraWan.cpp]: https://gist.github.com/xsleonard/7341172 that looks ... interesting.

[the remark this answers is not in the document]
Comment (rexfue) [reply multigeiger/loraWan.cpp]: interesting - yeah. But he uses malloc. I don't like malloc. You have to free memory after that. And i don't know, why he puts a '\0' at the end - it's a byte array, no string - isn't it?

Concerning python: rewrite the project using micro python ! ;-)

File: multigeiger/multigeiger.ino (hunk 7 above)
Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: usually when dealing with user input one needs some validation step (and rejection if invalid, letting the user try again).
and as the validation has to deal with types and valid values/ranges anyway, it usually can transform the string to the desired type also.

File: multigeiger/webconf.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/webconf.cpp]: the value is to have them in one place in a header file (and also to keep existing docs that describe this stuff).

File: multigeiger/webconf.cpp (hunk 4 above)
Comment (ThomasWaldmann) [reply multigeiger/webconf.cpp]: no, it is not better if it does not add information - you have to read all the stuff, it produces longer source code.

sometimes it also has typos or sounds strange...

File: multigeiger/webconf.cpp (hunk 9 above)
Comment (ThomasWaldmann) [reply multigeiger/webconf.cpp]: The `init` method does that via `loadConfig`.

I also found that one can set an own FormValidator.

File: multigeiger/multigeiger.ino (hunk 7 above)
Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: most codebases i have seen yet do a one-time read, validate, convert step when reading their text-based config (usually a text file in some format) and after that only use the appropriate types (bool, int, string, ...) but not the config text any more.

in our case here, the iotWebConf variables have to be considered our "text config" and the additional set of appropriately typed variables would contain the stuff we use from the loop.

if we immediately validate at the form input step and reject any form that has invalid data, we can immediately get the user to edit and resubmit the form, until we get input that validates.

Comment (rexfue) [reply multigeiger/multigeiger.ino]: OK, sounds good. Would you please try that?

File: multigeiger/webconf.cpp (hunk 9 above)
Comment (rexfue) [reply multigeiger/webconf.cpp]: Do you have another version of iotwebconf? There is no loadConfig. There is a configload, but that is private.

Comment (ThomasWaldmann) [reply multigeiger/webconf.cpp]: https://github.com/prampec/IotWebConf/blob/master/src/IotWebConf.cpp#L116

yeah, configLoad is private, but init is public.

File: multigeiger/utils.cpp [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    data[count] = strtol(buf, &endptr, 0);
+    pos += 2 * sizeof(char);
+    if (endptr[0] != '\0') {
+      //non-hexadecimal character encountered
+      return -1;
+    }
+  }
+  if (reverse) {
+    reverseByteArray(data, len);
+  }
```
Comment (ThomasWaldmann) [inline multigeiger/utils.cpp]: removing this would be cleaner. just call both functions from the higher layer.

File: multigeiger/utils.h [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+#ifndef UTILS_H
+#define UTILS_H
+
+// Prototypes
+int hex2data(unsigned char *data, const char *hexstring, unsigned int len, bool reverse);
+void reverseByteArray(unsigned char *data, int len);
+
+
+#endif
```
Comment (ThomasWaldmann) [inline multigeiger/utils.h]: add a linefeed at the end so diff does not complain.

File: multigeiger/webconf.cpp [hunk 12]
```diff
@@ -7,41 +7,40 @@
 #include <HTTPClient.h>
 #include "userdefines.h"
 
-// All these flags are ON (=1) at startup
-// They all can be changed at any time using the web interface
-char speakerTick[2] = "1";
-char startSound[2] = "1";
-char ledTick[2] = "1";
-char showDisplay[2] = "1";
+bool speakerTick = (SPEAKER_TICK == 1);
```
Comment (ThomasWaldmann) [inline multigeiger/webconf.cpp]: have SPEAKER_TICK as a boolean so this expression is not needed?

File: multigeiger/webconf.cpp [hunk 13]
```diff
@@ -105,9 +104,38 @@ void handleRoot(void) {  // Handle web requests to "/" path.
...
+
+void loadConfigVariables(void) {
+  // check if WiFi SSID has changed. If so, restart cpu. Otherwise, the program will not use the new SSID
+  if ((strcmp(lastWiFiSSID, "") != 0) && (strcmp(lastWiFiSSID, iotWebConf.getWifiSsidParameter()->valueBuffer) != 0)) {
+    log(INFO, "Doing restart...");
+    ESP.restart();
+  }
+  strcpy(lastWiFiSSID, iotWebConf.getWifiSsidParameter()->valueBuffer);
+
+  speakerTick = strcmp(speakerTick_c, "1") == 0;
```
Comment (ThomasWaldmann) [inline multigeiger/webconf.cpp]: ok for now, i'ld add something a bit better later.

Comment (rexfue) [reply multigeiger/webconf.cpp]: OK

File: multigeiger/webconf.cpp (hunk 12 above)
Comment (rexfue) [reply multigeiger/webconf.cpp]: ?? SPEAKER_TICK is a #define, so it is an integer , isn't it?

File: multigeiger/utils.cpp [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+  return 0;
+}
+
+// reverse a bytearray of length len
```
Comment (ThomasWaldmann) [inline multigeiger/utils.cpp]: "only works for even lengths"

File: multigeiger/utils.h [hunk 15]
```diff
@@ -0,0 +1,11 @@
+// Div. utilities
```
Comment (ThomasWaldmann) [inline multigeiger/utils.h]: see above.

File: multigeiger/webconf.cpp (hunk 12 above)
Comment (ThomasWaldmann) [reply multigeiger/webconf.cpp]: a #define is a preprocessor text replacement, so it can be a lot of things.

File: multigeiger/webconf.cpp [hunk 16]
```diff
@@ -90,3 +185,6 @@ void setup_webconf() {
   });
 }
 
+
+
+
```
Comment (ThomasWaldmann) [inline multigeiger/webconf.cpp]: remove?

File: multigeiger/webconf.h [hunk 17]
```diff
@@ -6,9 +6,24 @@
...
+extern char appeui[];
+extern char deveui[];
+extern char appkey[];
+#endif
+
 extern char ssid[];
 extern IotWebConf iotWebConf;
 
 void setup_webconf(void);
+boolean parameterTrue(char *parameter);
```
Comment (ThomasWaldmann) [inline multigeiger/webconf.h]: remove?

Comment (rexfue) [reply multigeiger/webconf.h]: OK

File: multigeiger/webconf.cpp (hunk 16 above)
Comment (rexfue) [reply multigeiger/webconf.cpp]: OK

File: multigeiger/utils.cpp (hunk 14 above)
Comment (rexfue) [reply multigeiger/utils.cpp]: comment added

File: multigeiger/utils.cpp [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+#include "utils.h"
+
+// ** convert hexstring to len bytes of data
+//returns 0 on success, -1 on error
+//data is a buffer of at least len bytes
+//hexstring is upper or lower case hexadecimal, NOT prepended with "0x"
+//convert hexstring to len bytes of data
+//returns 0 on success, -1 on error
+//data is a buffer of at least len bytes
+//hexstring is upper or lower case hexadecimal, NOT prepended with "0x"
```
[the remark this answers is not in the document]
Comment (rexfue) [reply multigeiger/utils.cpp]: The style checker didn't complain :-(

File: multigeiger/utils.cpp [hunk 19]
```diff
@@ -0,0 +1,49 @@
+// Div utilities
```
[the remark this answers is not in the document]
Comment (rexfue) [reply multigeiger/utils.cpp]: Changed to misc

File: multigeiger/transmission.cpp [hunk 20]
```diff
@@ -221,32 +216,34 @@ void transmit_data(String tube_type, int tube_nbr, unsigned int dt, unsigned int
...
-  #endif
 
-  #if SEND2LORA
-  log(INFO, "Sending to TTN ...");
-  displayStatusLine("TTN");
-  send_ttn_geiger(tube_nbr, dt, gm_counts);
-  if (have_thp)
-    send_ttn_thp(temperature, humidity, pressure);
+  #if CPU==STICK
+  if(sendToLora && (strcmp(appeui,"") != 0)) { // send only, if we have LoRa credentials
```
[the remark this answers is not in the document]
Comment (rexfue) [reply multigeiger/transmission.cpp]: No, this style checker doesn't work well, i guess.

File: multigeiger/loraWan.cpp [hunk 21]
```diff
@@ -48,10 +46,14 @@
...
 #include "hal/heltecv2.h"
 #include <SPI.h>
+#include "webconf.h"
 
 #include "loraWan.h"
 #include "log.h"
 
+// Prototype
+int hex2data(unsigned char *data, const char *hexstring, unsigned int len, bool reverse);
+
```
[the remark this answers is not in the document]
Comment (rexfue) [reply multigeiger/loraWan.cpp]: OK

File: multigeiger/loraWan.cpp [hunk 22]
```diff
@@ -16,14 +19,9 @@
 // DIO2 (10) = GPIO32 (12)
 
 #include "userdefines.h"
+#include "webconf.h"
+#include "utils.h";
```
[the remark this answers is not in the document]
Comment (rexfue) [reply multigeiger/loraWan.cpp]: Uhps

**Comment by rexfue** [comment]
Sorry, commented already 6 days ago but forgot to click on comment. PR is out

File: multigeiger/transmission.cpp [hunk 23]
```diff
@@ -221,32 +216,34 @@ void transmit_data(String tube_type, int tube_nbr, unsigned int dt, unsigned int
...
-  #if SEND2SENSORCOMMUNITY
-  log(INFO, "Sending to sensor.community ...");
-  displayStatusLine("sensor.community");
-  send_http_geiger(SENSORCOMMUNITY, dt, hv_pulses, gm_counts, cpm, XPIN_RADIATION);
-  if (have_thp) {
-    send_http_thp(SENSORCOMMUNITY, temperature, humidity, pressure, XPIN_BME280);
+  if(sendToCommunity) {
+    log(INFO, "Sending to sensor.community ...");
+    displayStatusLine("sensor.community");
+    send_http_geiger(SENSORCOMMUNITY, dt, hv_pulses, gm_counts, cpm, XPIN_RADIATION)  ;
```
Comment (ThomasWaldmann) [inline multigeiger/transmission.cpp]: remove the blanks before the semicolon, please.

File: multigeiger/utils.cpp [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+// returns 0 on success, -1 on error
+// data is a buffer of at least len bytes
+// hexstring is upper or lower case hexadecimal, NOT prepended with "0x"
+int hex2data(unsigned char *data, const char *hexstring, unsigned int len) {
+  const char *pos = hexstring;
+  char *endptr;
+  size_t count = 0;
+  if ((hexstring[0] == '\0') || (strlen(hexstring) % 2)) {
+    //hexstring contains no data
+    //or hexstring has an odd length
```
Comment (ThomasWaldmann) [inline multigeiger/utils.cpp]: // + 1 blank

Comment (ThomasWaldmann) [reply multigeiger/utils.cpp]: (can you search and replace that globally?)

**Comment by rexfue** [comment]
Fixed. Can you please make the 1.13 release after merging. Thanks.

**Comment by rexfue** [comment]
Merged. You can make the 1.13 release.

**Pull request merged**